File: 20. Valid Parentheses/solution.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    if(s.length % 2 === 1) {
        return false
    }
    let stack = []
    let map = {
        '{': '}',
        '(': ')',
        '[': ']'
    }
    for(let c of s) {
        if(!stack.length) {
            stack.push(c)
        } else {
            let top = stack[stack.length - 1]
            if(map[top] === c) {
                stack.pop()
            } else {
                stack.push(c)
            }
        }
    }
    return !Boolean(stack.length) 
};

File: 20. Valid Parentheses/solution.go
func isValid(s string) bool {
    if len(s) == 0 {
        return true
    }
    if len(s) % 2 == 1 {
        return false
    }
    stack := []string{string(s[0])}
    parenthesesMap := map[string]string{"(": ")", "{": "}", "[": "]"}
    for i:=1; i<len(s); i++ {
        if len(stack) == 0 {
            stack = append(stack, string(s[i]))
            continue
        }
        if parenthesesMap[stack[len(stack)-1]] == string(s[i]) {
            stack = stack[:len(stack)-1]
        } else {
            stack = append(stack, string(s[i]))
        }
    }
    if len(stack) == 0 {
        return true
    } else {
        return false
    }
}